### PR TITLE
fix(config): always merge default allowed_remote_resources into loaded configs

### PR DESCRIPTION
**Pull request by ggallen**
## Summary

- Add `MergeDefaultAllowedRemoteResources` to merge the two first-party URL prefixes (`fullsend-ai/fullsend` and `fullsend-ai/agents`) into every loaded config at the `orgconfig.go` loading layer
- When no config file exists, `run.go` and `lock.go` now initialize `orgAllowlist` to the defaults instead of nil
- Remove the narrow `propagateDefaultAllowlist` fix from PR #3425, which is superseded by this broader approach

## Context

Repos that omit `allowed_remote_resources` from their `config.yaml` (like `fullsend-playground/node-api`) hit `"URL not in allowed_remote_resources"` errors when the harness composition step tries to resolve relative resources (pre/post scripts) from URL-sourced harnesses. PR #3425 patched this narrowly on the agents-repo fallback path, but the root fix is to ensure the defaults are always present after config loading.

Failing run: https://github.com/fullsend-playground/node-api/actions/runs/28953629888/job/85906703438?pr=6
Related: #3396

## Test plan

- [x] `TestMergeDefaultAllowedRemoteResources` — 7 subtests covering nil, empty, custom, dedup, partial overlap, idempotency, no-mutation
- [x] Updated existing orgconfig loading tests to expect defaults merged in
- [x] `go test ./internal/config/... ./internal/cli/...` passes
- [x] `go vet ./...` clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/cli/orgconfig.go [hunk 1]
```diff
@@ -61,13 +61,16 @@ func tryLoadFullsendConfig(path string, printer *ui.Printer) *config.OrgConfig {
 			printer.StepWarn("Per-repo config malformed (remote resource allowlist unavailable): " + perRepoErr.Error())
 			return nil
 		}
-		return config.OrgConfigFromPerRepo(perRepo)
+		orgCfg := config.OrgConfigFromPerRepo(perRepo)
+		orgCfg.AllowedRemoteResources = config.MergeDefaultAllowedRemoteResources(orgCfg.AllowedRemoteResources)
```
Comment (waynesun09) [inline internal/cli/orgconfig.go]: **[HIGH] Unconditional trust widening, inconsistently applied across the codebase**

This merge is applied unconditionally to every successfully parsed config — including configs that already declare a custom, deliberately narrow allowlist that doesn't include the fullsend-ai prefixes. Every org now unconditionally trusts `raw.githubusercontent.com/fullsend-ai/*` regardless of what they configured, widening the supply-chain trust boundary beyond what the PR description implies ("fixing configs that omit the field").

This is also inconsistently enforced: `internal/cli/agent.go`/`migrate.go` (config-editing paths) and several `internal/cli/admin.go`/`internal/layers/*.go` read paths (repo enable/disable, uninstall, health checks) parse `config.yaml` directly via `config.ParseOrgConfig` and never call this merge — so "always merge" only holds for the `run`/`lock` execution path, not the codebase as a whole.

**Suggestion:** Confirm with stakeholders whether widening trust for orgs with existing custom allowlists is intentional; if so, document it explicitly. Consider centralizing the merge inside `ParseOrgConfig`/`ParsePerRepoConfig` so every caller gets consistent behavior, or explicitly scope+document which paths are exempt and why.

File: internal/cli/run.go [hunk 2]
```diff
@@ -228,7 +228,7 @@ func runAgent(ctx context.Context, agentName, fullsendDir, outputBase, targetRep
 	// HasURLReferences will enforce its presence later if needed.
 	orgConfigPath := filepath.Join(absFullsendDir, "config.yaml")
 	orgCfg := tryLoadOrgConfig(orgConfigPath, printer)
-	var orgAllowlist []string
+	orgAllowlist := config.DefaultAllowedRemoteResources()
```
Comment (waynesun09) [inline internal/cli/run.go:309]: **[MEDIUM] Inconsistent allowlist-defaulting logic duplicated across four sites**

There are now at least three separate places implementing "default to `config.DefaultAllowedRemoteResources()` when nothing is configured": this `orgCfg==nil` fallback here (and the equivalent in `lock.go:177`), the merge inside `orgconfig.go`, and `tryAgentsRepoFallback`'s own nil-check (`run.go:2754-2757`, now dead code on this path since `composeOpts.OrgAllowlist` can no longer be nil by the time it's called from here). This isn't currently broken but is a maintainability hazard — one implementation could be updated without the others, silently reintroducing the class of bug this PR fixes.

**Suggestion:** Consolidate into a single helper function that is the sole source of truth for "the effective allowlist," and remove the now-dead nil-check in `tryAgentsRepoFallback` (or comment clearly why it's retained as defense-in-depth).

File: internal/config/config.go [hunk 3]
```diff
@@ -179,6 +179,29 @@ func DefaultAllowedRemoteResources() []string {
...
+	if len(existing) == 0 {
+		return defaults
+	}
+	seen := make(map[string]bool, len(defaults))
+	for _, d := range defaults {
+		seen[d] = true
+	}
+	result := make([]string, len(defaults))
+	copy(result, defaults)
+	for _, e := range existing {
```
Comment (waynesun09) [inline internal/config/config.go:203]: **[MEDIUM] Merge always prepends defaults, changing prefix-match precedence and audit-log attribution**

The merge places the two default prefixes first regardless of the original config's ordering. Prefix-matching lookups (e.g. `MatchingAllowedPrefixInList`) return the *first* matching prefix, and ADR-0038's audit log schema records "which prefix matched" (`allowed_by`). If an operator's own custom entry happens to overlap a default's URL space, this reordering changes which prefix gets reported as the match, altering audit trail semantics.

**Suggestion:** Preserve the original list's order and append only missing defaults at the end, rather than always prepending defaults, unless order is confirmed to have no functional significance beyond dedup.

File: internal/config/config_test.go [hunk 4]
```diff
@@ -1704,3 +1704,49 @@ func TestOrgConfigFromPerRepo(t *testing.T) {
...
+
+	t.Run("nil input returns defaults", func(t *testing.T) {
+		result := MergeDefaultAllowedRemoteResources(nil)
+		assert.Equal(t, defaults, result)
+	})
+
+	t.Run("empty input returns defaults", func(t *testing.T) {
+		result := MergeDefaultAllowedRemoteResources([]string{})
+		assert.Equal(t, defaults, result)
+	})
```
Comment (waynesun09) [inline internal/config/config_test.go:1757]: **[MEDIUM] Test asserts current behavior tautologically instead of validating the actual security question**

The nil and empty-input subtests here both assert `defaults` is returned — this merely restates the `len(existing) == 0` line verbatim rather than testing whether nil and explicit-empty *should* be treated the same (the crux of the deny-all regression flagged elsewhere in this PR).

Separately, several `run_test.go` assertions (e.g. lines 278, 357, 405) build expected values via `append(config.DefaultAllowedRemoteResources(), "https://example.com/")` — safe today only because `DefaultAllowedRemoteResources()` returns a fresh 2-element literal each call; this is a footgun if that implementation ever changes to a slice with spare capacity.

**Suggestion:** Add a test that documents/justifies the nil-vs-empty decision explicitly rather than asserting current behavior tautologically. Replace `append`-on-return-value patterns with explicit literal construction in tests.

File: internal/cli/run_test.go [hunk 5]
```diff
@@ -3555,44 +3547,21 @@ post_script: scripts/post-triage.sh
...
-
-	// Override allowlist to match test server (default points at github.com).
-	opts.OrgAllowlist = []string{srv.URL + "/"}
 
 	h, _, err := harness.LoadWithBase(context.Background(), harnessPath, opts)
-	require.NoError(t, err, "should succeed with allowlist propagated")
+	require.NoError(t, err, "should succeed with allowlist set")
 	assert.True(t, filepath.IsAbs(h.PreScript), "pre_script should be resolved to cache path")
 	assert.True(t, filepath.IsAbs(h.PostScript), "post_script should be resolved to cache path")
 }
```
Comment (waynesun09) [inline internal/cli/run_test.go:3774]: **[MEDIUM] No integration test proves the fix against the original failure mode**

`TestAgentsRepoFallback_LoadWithBase_DefaultAllowlist` manually constructs `OrgAllowlist: []string{srv.URL + "/"}` and calls `harness.LoadWithBase` directly — it never exercises the actual `runAgent` entry point where `orgCfg == nil` → `orgAllowlist := config.DefaultAllowedRemoteResources()` → `tryAgentsRepoFallback` → `LoadWithBase`. No test simulates "no config.yaml file at all," which is the exact scenario from the linked failing run (`fullsend-playground/node-api`).

The negative-case assertion from the old test (nil allowlist + SourceURL → `"not in allowed_remote_resources"` error) was also dropped, weakening regression coverage for the harness package's own deny-by-default behavior.

**Suggestion:** Add an end-to-end test through `runAgent` (or minimally the full `orgCfg==nil` → default → fallback → `LoadWithBase` chain) that doesn't pre-populate the allowlist manually. Restore the negative-case assertion alongside the positive one.

**Review by waynesun09 — COMMENTED** [review]
Review squad (5 agents: Claude ×2, Gemini, Codex, Claude) unanimously flagged a CRITICAL deny-all regression already covered by existing automated review comments on `config.go:188/190` (not re-posted here to avoid duplication). Posting 5 additional unique MEDIUM+ findings not covered elsewhere: unconditional/inconsistent trust widening, missing integration test coverage for the actual reported failure, duplicated allowlist-defaulting logic across 4 call sites, allowlist reordering affecting audit-log attribution, and tautological test coverage.

File: internal/config/config.go [hunk 6]
```diff
@@ -179,6 +179,29 @@ func DefaultAllowedRemoteResources() []string {
...
 
+// MergeDefaultAllowedRemoteResources returns a new slice containing the
+// default allowed-remote-resources prefixes merged with the provided list.
+// Entries already present are not duplicated. Returns defaults alone for
+// nil or empty input.
+func MergeDefaultAllowedRemoteResources(existing []string) []string {
+	defaults := DefaultAllowedRemoteResources()
+	if len(existing) == 0 {
+		return defaults
+	}
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/config/config.go]: Fixed. Changed `len(existing) == 0` to `existing == nil` so that explicit `allowed_remote_resources: []` preserves deny-all semantics. Also renamed function to `EnsureDefaultAllowedRemoteResources` and reversed the ordering to append defaults instead of prepending.

File: internal/config/config.go [hunk 7]
```diff
@@ -179,6 +179,29 @@ func DefaultAllowedRemoteResources() []string {
 	}
 }
 
+// MergeDefaultAllowedRemoteResources returns a new slice containing the
+// default allowed-remote-resources prefixes merged with the provided list.
+// Entries already present are not duplicated. Returns defaults alone for
+// nil or empty input.
+func MergeDefaultAllowedRemoteResources(existing []string) []string {
+	defaults := DefaultAllowedRemoteResources()
+	if len(existing) == 0 {
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/config/config.go]: Fixed. Changed guard from `len(existing) == 0` to `existing == nil`. Nil (field omitted from YAML) gets defaults; explicit empty `[]string{}` is returned unchanged to preserve deny-all.

File: internal/cli/run.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/run.go:309]: Acknowledged. This is centralization of existing behavior — the old `propagateDefaultAllowlist` achieved the same effective result through the agents-repo-fallback path. No new permission grant.

File: internal/config/config.go [hunk 8]
```diff
@@ -179,6 +179,29 @@ func DefaultAllowedRemoteResources() []string {
 	}
 }
 
+// MergeDefaultAllowedRemoteResources returns a new slice containing the
+// default allowed-remote-resources prefixes merged with the provided list.
+// Entries already present are not duplicated. Returns defaults alone for
+// nil or empty input.
+func MergeDefaultAllowedRemoteResources(existing []string) []string {
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/config/config.go]: Fixed. Renamed to `EnsureDefaultAllowedRemoteResources` to clarify pure-function semantics (returns a new slice, does not modify in place).

File: internal/cli/orgconfig.go (hunk 1 above)
Comment (ggallen) [reply internal/cli/orgconfig.go]: Fixed the nil-vs-empty conflation. With the `existing == nil` guard, an explicit `allowed_remote_resources: []` now preserves deny-all. Custom allowlists that omit first-party prefixes get defaults appended — this is intentional so base composition and agents-repo fallback work for all orgs. The inconsistency across `agent.go`/`migrate.go`/`admin.go` paths is by design — those are config-editing paths, not execution paths that resolve remote resources.

File: internal/cli/run.go (hunk 2 above)
Comment (ggallen) [reply internal/cli/run.go:309]: Fixed. Removed the dead nil-check in `tryAgentsRepoFallback` (lines 2754-2757) since callers now always set the allowlist. The remaining two defaulting sites (`run.go` and `lock.go` for when `orgCfg == nil`) serve a different code path from `EnsureDefaultAllowedRemoteResources` in `orgconfig.go` (which handles parsed configs).

File: internal/config/config.go (hunk 3 above)
Comment (ggallen) [reply internal/config/config.go:203]: Fixed. Reversed the ordering — the caller's original entries are preserved first, and only missing defaults are appended at the end. This preserves prefix-match precedence and audit-log attribution.

File: internal/config/config_test.go (hunk 4 above)
Comment (ggallen) [reply internal/config/config_test.go:1757]: Fixed. Tests now explicitly validate nil-vs-empty semantics: nil returns defaults, explicit empty preserves deny-all. Replaced all `append(config.DefaultAllowedRemoteResources(), ...)` patterns with explicit literal construction to avoid the spare-capacity footgun.

File: internal/cli/run_test.go (hunk 5 above)
Comment (ggallen) [reply internal/cli/run_test.go:3774]: Added loader-level tests: `TestTryLoadFullsendConfig_ExplicitEmptyAllowlist`, `TestTryLoadFullsendConfig_OmittedAllowlist`, and `TestRequireFullsendConfig_ExplicitEmptyAllowlist` that exercise the full config-loading path with explicit empty and omitted `allowed_remote_resources` fields.

**Review by ralphbean — APPROVED** [review]
We talked in slack about the deny-all fail-open problem and agreed that we're just going to accept it here and merge this to get people unbroken. We'll return shortly in the future and remove it.

**Review by waynesun09 — APPROVED** [review]
Reviewed by a 4-agent squad (2x Claude, Gemini, Codex/Cursor). No blocking issues — the core fix (centralizing default allowlist merge at the orgconfig loading layer, replacing the narrow #3425 patch) is correct, and the deny-all (`allowed_remote_resources: []`) contract is preserved and well-tested.

A few non-blocking follow-ups worth considering:
- `EnsureDefaultAllowedRemoteResources` unconditionally appends the first-party prefixes to any non-empty custom allowlist with no opt-out short of deny-all. This is a reasonable trade-off (the tool needs these URLs for harness composition) but is a real trust-boundary change worth a docs/changelog note beyond the inline YAML comment.
- `internal/cli/agent.go`'s `loadAgentConfig` (used by `fullsend agent add/list/update/remove`) calls `config.ParseOrgConfig`/`ParsePerRepoConfig` directly and doesn't go through `EnsureDefaultAllowedRemoteResources`, so `fullsend agent add <url>` on a config that omits `allowed_remote_resources` could still hit a misleading "not covered by allowed_remote_resources" validation error.
- `TestRequireFullsendConfig_OmittedAllowlist` is missing as a counterpart to `TestTryLoadFullsendConfig_OmittedAllowlist` — both loaders independently call the merge function.

None of these block merging; approving.

File: docs/guides/user/customizing-agents.md [hunk 9]
```diff
@@ -64,7 +64,8 @@ validation_loop:                     # script is required; these sub-fields are
   feedback_mode: stderr          # "stderr", "stdout", or "exit_code" (optional)
 
 allowed_remote_resources:        # URL prefixes allowed for remote skills/agents/policies
-  - https://github.com/org/       # Resources must match a prefix to be fetched
+  - https://github.com/org/       # First-party defaults are auto-appended when entries exist
```
[the remark this answers is not in the document]
Comment (ggallen) [reply docs/guides/user/customizing-agents.md]: Fixed. Updated the inline docs to describe all three behaviors: omitted field gets first-party defaults, non-empty list gets defaults appended, explicit `[]` preserves deny-all.

File: internal/cli/run.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/run.go:309]: Fixed. Added comment clarifying that this default covers the absent-config case, while `EnsureDefaultAllowedRemoteResources` in `orgconfig.go` covers the omitted-field case when a config is present.

File: internal/cli/lock.go [hunk 10]
```diff
@@ -173,7 +174,7 @@ func lockOneAgent(ctx context.Context, agentName, absFullsendDir, forgeFlag stri
 
 	orgConfigPath := filepath.Join(absFullsendDir, "config.yaml")
 	orgCfg := tryLoadOrgConfig(orgConfigPath, printer)
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/lock.go:176]: Fixed. Added same clarifying comment as in `run.go` — this default covers absent config, `EnsureDefaultAllowedRemoteResources` covers the omitted-field case in loaded configs.

File: internal/config/config.go [hunk 11]
```diff
@@ -179,6 +179,34 @@ func DefaultAllowedRemoteResources() []string {
 	}
 }
 
+// EnsureDefaultAllowedRemoteResources returns a new slice with default
+// allowed-remote-resources prefixes merged into the provided list.
+// Nil input (field omitted from YAML) returns defaults alone.
+// An explicit empty slice (allowed_remote_resources: []) is returned
+// unchanged to preserve deny-all semantics. Non-empty input gets any
+// missing defaults appended (preserving the caller's original ordering).
+func EnsureDefaultAllowedRemoteResources(existing []string) []string {
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/config/config.go:194]: Acknowledged — intentional design trade-off, unchanged across 4 review cycles. First-party defaults are always included so base composition and agents-repo fallback work for all orgs. The deny-all escape hatch (`[]`) is preserved and documented.

File: docs/guides/user/customizing-agents.md [hunk 12]
```diff
@@ -64,7 +64,9 @@ validation_loop:                     # script is required; these sub-fields are
   feedback_mode: stderr          # "stderr", "stdout", or "exit_code" (optional)
 
 allowed_remote_resources:        # URL prefixes allowed for remote skills/agents/policies
```
[the remark this answers is not in the document]
Comment (ggallen) [reply docs/guides/user/customizing-agents.md:66]: Acknowledged. The inline YAML comments now describe all three behaviors. The specific prefixes (`fullsend-ai/fullsend` and `fullsend-ai/agents`) are implementation details documented in the function godoc and not appropriate for user-facing config examples.

File: internal/cli/run.go [hunk 13]
```diff
@@ -228,7 +228,9 @@ func runAgent(ctx context.Context, agentName, fullsendDir, outputBase, targetRep
 	// HasURLReferences will enforce its presence later if needed.
 	orgConfigPath := filepath.Join(absFullsendDir, "config.yaml")
 	orgCfg := tryLoadOrgConfig(orgConfigPath, printer)
-	var orgAllowlist []string
+	// Fallback for absent config; EnsureDefaultAllowedRemoteResources
+	// handles the omitted-field case when a config is present.
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/run.go:308]: Already addressed — clarifying comment was added in the prior push. The bot confirms "the comment at the initialization site adequately explains the dual mechanism."

File: internal/config/config.go [hunk 14]
```diff
@@ -185,6 +185,34 @@ func DefaultAllowedRemoteResources() []string {
 	}
 }
 
+// EnsureDefaultAllowedRemoteResources returns a new slice with default
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/config/config.go:188]: Acknowledged — same design trade-off, stable across 5 review cycles. First-party defaults are always included unless deny-all (`[]`) is used. Documented in customizing-agents.md.

File: internal/cli/run.go [hunk 15]
```diff
@@ -304,7 +304,9 @@ func runAgent(ctx context.Context, agentName, fullsendDir, outputBase, targetRep
 	// HasURLReferences will enforce its presence later if needed.
 	orgConfigPath := filepath.Join(absFullsendDir, "config.yaml")
 	orgCfg := tryLoadOrgConfig(orgConfigPath, printer)
-	var orgAllowlist []string
+	// Fallback for absent config; EnsureDefaultAllowedRemoteResources
```
[the remark this answers is not in the document]
Comment (ggallen) [reply internal/cli/run.go:307]: Already addressed — clarifying comment at the initialization site explains the dual mechanism (absent config vs. present config with omitted field).
